### ISSUE TITLE
Handler for the spot termination signals

### ISSUE THREAD

**Issue by cristim** [AutoSpotting/AutoSpotting#133]
# Github issue

## Issue type ##

- Feature Idea

## Summary ##

Autospotting should have a way to proactively react on terminating spot instances as soon as they are flagged for termination.

## Detailed implementation description

Each instance should optionally have injected into the user_data script a component continuously monitoring the spot termination metadata entry. This could also be injected by the user if desired, but I like these things to just work with no configuration, consistent to the other AutoSpotting features which also don't require much if any configuration.

When the spot instance is flagged for termination, it should be able to notify AutoSpotting immediately, in order to launch a replacement spot instance or maybe even fall back to an on-demand instance, but more importantly to cleanly detach the spot instance from the group before it is terminated.

The least intrusive way I could think of, which doesn't require additional IAM permissions to the instance profile, would be to introduce an API Gateway REST endpoint which would accept POST requests containing the spot instance ID. The instance would perform a request on the API Gateway to give autospotting some information about itself, such as instance ID, availability zone and instance type, maybe more if needed to make the replacement logic as simple as possible. 

Once such a request is received by AutoSpotting, it would immediately detach the spot instance with that ID from its group, reducing the capacity, and launch a new compatible spot instance in the same availability zone, in order to restore capacity. In case another spot instance can be launched immediately and the startup time is short, the capacity drop would be brief and AutoScaling actions may not even be triggered so the transition between the spot instances could be done without any on-demand instance being launched. If there is an on-demand instance, the normal replacement logic would be executed later when the spot instance is out of the grace period, otherwise the spot instance is simply added to the group, restoring capacity to the previous value.

The API Gateway endpoint would be configured automatically by the CloudFormation/Terraform infrastructure code and the endpoint would be passed to the AutoSpotting Lambda function as another more or less hardcoded environment variable.

## Credits
Many thanks to @arunasank, I got this idea while watching her awesome [talk](https://www.youtube.com/watch?v=XQJ7YhVoSWI) about [spotswap](https://github.com/mapbox/spotswap)

**Comment by erolosty** [comment]
Wow, this would be incredible

**Comment by otisg** [comment]
That would indeed be awesome!

**Comment by roeyazroel** [comment]
@cristim currently the "Spot Instance interruption notices" is two minutes, in order to replace the spot with a spot, the action should be in one shot - "drain + remove from asg + request a new one + add it to the asg". this will cause the function to be executed more than 5 minutes, unless the autospotting function will be executed every 30 seconds...

**Comment by cristim** [comment]
It would be drain and remove with capacity temporarily decreased, when above the minimum, and launch a new one which will be attached back later once passed the grace period, but without terminating or detaching a corresponding on demand instance.

The group's scaling policy may in the meantime launch an on demand instance in the same availability zone, which would be replaced by the currently implemented logic.

If the scaling isn't happening, the group will run with decreased capacity until the new spot instance is ready.

I don't think this is problematic, it should be doable with reasonable effort.

**Comment by roeyazroel** [comment]
In case it will be relay on the ASG on demand instance that will be automatically replace, I agree.

**Comment by vecchp** [comment]
Now that Amazon is doing per second billing for EC2 instances is replacing a marked for termination instance with a spot very critical?  It would seem better to fall back to the on-demand and let it replace later.  It could simplify the logic by just removing instances marked for termination from the ASG.

**Comment by cristim** [comment]
I agree with @vecchp, there are just some relatively small benefits:
- the spot instances can be properly drained
- avoiding the on demand instances entirely will migrate the traffic only once to the new spot instance, instead of twice, so it is a bit less disruptive.

**Comment by xlr-8** [comment]
@cristim could you update the description with the end task? I believe @vecchp & you summarized the new need pretty well.

**Comment by cristim** [comment]
@xlr-8 please explain how you think it would look like, I have a hard time articulating it further.

Or feel free to update it as you see fit, as maintainer you should also have the required permissions.

**Comment by xlr-8** [comment]
As recently announced by AWS those events are now available from Cloudwatch: https://aws.amazon.com/about-aws/whats-new/2018/01/amazon-ec2-spot-two-minute-warning-is-now-available-via-amazon-cloudwatch-events/

**Comment by cristim** [comment]
Thanks @xlr-8, that's a huge step forward, as part of the AutoSpotting stack we can create such a rule and an SNS topic that runs a second Lambda function handling the termination signals. Then slowly separate the code into multiple such functions to simplify the logic and make it more modular.


no longer need

**Comment by cristim** [comment]
I'm contemplating to have this feature a proprietary additional module  offered as a bundle to the users who get the stable builds.

Would you be interested in paying for this kind of bundle? If so, how much per month/year? What about a percentage of the savings? How much would you find reasonable?

**Comment by CpuID** [comment]
> I'm contemplating to have this feature a proprietary additional module offered as a bundle to the users who get the stable builds.

................................

Apologies, but no. I understand you are trying to move to a model that allows you to support your development time, but I personally disagree with this particular route (freemium). I have seen Hashicorp do it with numerous features in their stack, and all it has done is prompted me to consider alternative technologies for fear of "premium creep" (more features being paywalled, eventually covering one I can't do without). This also turned me off using InfluxDB (paid clustering) along the way, in lieu of other tech.

I am a big believer in open source, and others contributing back their changes made for the greater community good. I feel the approach considered would be a step backwards from this (greys the waters between open source contributions and whose getting paid for what works etc).

**Comment by CpuID** [comment]
> I'm contemplating to have this feature a proprietary additional module offered as a bundle to the users who get the stable builds.

In addition, this has prompted me in the past to "reinvent the wheel" for simple functionality and just put a DIY workaround in place to achieve the same result.

**Comment by cristim** [comment]
Thanks @CpuID, this is very valuable feedback, I really appreciate it.

I spent countless hours working on this project without expecting anything in return, simply because I liked the idea and I thought it should be implemented and I enjoyed working on it. I also got a lot of help from various contributors whom I owe a lot, this project wouldn't be as advanced without their help, so it's not fair for me to put paywalls for any code I didn't write.

I think this project can make a huge difference and I'd like to continue working on it even more than before, but for personal reasons I can no longer afford to invest a significant amount of time unless I get something in return. I also can't support the AWS costs for the infrastructure that powers the easy installation method and the website from my own pocket forever, so that's why I was thinking on ways to make all this sustainable. This closed source feature was just such an attempt, the other being to ask the people who use the easy installation method to contribute a small fraction of their savings. 

Recently I started accepting such contributions on [Patreon](https://www.patreon.com/cristim), so people who want to support the infrastructure costs and further development can contribute there. Going forward I will keep developing in the open, but I'll allocate as much time as I can justify to myself and my family based on those contributions.

So far I have a single patron that contributes $5/month. I'm really thankful to him but unfortunately this is barely enough to cover my AWS costs, so I can't spend time much if any time doing feature development. I do my best to answer user issues and review pull requests, but this is already eating a lot of my time.

If anyone wants me to work on this feature (or any other for that matter) this is how they can make it happen... unless they want to spend the time and contribute it themselves.

**Comment by cristim** [comment]
And regarding this particular feature, I think as of the last year's Re:invent - when AWS introduced stable pricing for spot instances - this became largely a non-issue, since it rarely happens that instances are terminated nowadays.

**Comment by jjones-smug** [comment]
I have add-on Python Lambda that does this by subscribing to "EC2 Spot Interrupt" notifications in Cloudwatch Events.  The idea is the same as polling the metadata, but it's a "push" model instead.  You can setup CloudWatch Events to trigger a Lambda or add a record to a Kinesis stream.  We're using the Kinesis stream option, since the data is interesting to keep around.  So the Kinesis stream is setup to trigger the Lambda when the records is added.

In practice you wind up getting a little over 2 minutes warning before the spot instances it terminated.  So the ASG can at least gracefully shutdown the instance and launch an OD replacement in that window.  Even though in our case the new OD instance often isn't fully ready before the spot instance is terminated, it still helps.

Anyway, the point of this posted is to noted that you can trigger a Lambda function when the spot interrupts are published now, you don't have to poll.

**Comment by cristim** [comment]
We recently added support for this, further implementation for launching new spot instances should be raised in another issue.
